usaspending_api/references/management/commands/load_historical_appropriation_account_balances.py

```diff
@@ -0,0 +1,151 @@
+    "cars_dump__obligations__2020_03_17.csv": "temp_load_historical_file_a_data_act_obligations",
+    "cars_dump__outlays__2020_03_17.csv": "temp_load_historical_file_a_data_act_outlays",
+}
+
+# column_name: data_type
+DATA_TYPE_MAPPING = {
+    "ALLOC_XFER_AGENCY": "text",
+    "AGENCY_IDENTIFIER": "text",
+    "AVAIL_TYPE_CODE": "text",
+    "BEG_POA": "text",
+    "END_POA": "text",
+    "MAIN_ACCT": "text",
+    "SUB_ACCT": "text",
+    "TAS_TITLE": "text",
+    "FR_ENTITY": "text",
+    "BUDGET_SUBFUNCTION": "text",
+    "STMT_DETAIL_LABEL": "text",
+    "FY": "smallint",
+    "PD": "smallint",
+    "BUDGETARY_RESOURCES_AMOUNT": "decimal(23, 2)",
+    "DEOBLIGATIONS_AMOUNT": "decimal(23, 2)",
+    "OBLIGATIONS_AMOUNT": "decimal(23, 2)",
+    "OUTLAYS_AMOUNT": "decimal(23, 2)",
+}
+
+READ_BUFFER_SIZE_BYTES = 1024 * 1024
+
+
+class Timer(ScriptTimer):
+    """Make logging a little less verbose."""
+
+    def log_starting_message(self):
+        pass
+
+
+class Command(BaseCommand):
+    help = (
+        "One off loader for pulling in historical File A CARS data from CSV files provided by the "
+        "product owner.  Performs a full replace every run."
+    )
+
+    def add_arguments(self, parser):
+
+        parser.add_argument(
+            "source_location",
+            default=DEFAULT_CSV_LOCATION,
+            metavar="SOURCEURI",
+            nargs="?",
+            help=(
+                "Path or URI to the location of the raw CSV files to be loaded.  Raw files MUST BE NAMED "
+                f"{', '.join(FILE_TO_TABLE_MAPPING.keys())}!  Default location: {DEFAULT_CSV_LOCATION}"
+            ),
+        )
+
+    def handle(self, *args, **options):
+
+        source_location = options["source_location"]
+        logger.info(f"SOURCE CSV LOCATION: {source_location}")
+
+        with psycopg2.connect(get_database_dsn_string()) as connection:
+            with connection.cursor() as cursor:
+                self.connection = connection
+                self.cursor = cursor
+                for file_name, table_name in FILE_TO_TABLE_MAPPING.items():
+                    with Timer(f"Copy {file_name}"):
+                        uri = os.path.join(source_location, file_name)
+                        file_path = RetrieveFileFromUri(uri).copy_to_temporary_file()
+                    with Timer(f"Get CSV headers from {file_name}"):
+                        headers = self._get_headers(file_path)
+                    with Timer(f"Create temporary table {table_name}"):
+                        self._create_temporary_table(table_name, headers)
+                    with Timer(f"Import {file_name}"):
+                        self._import_file(file_path, table_name)
+                    os.remove(file_path)
+
+                destination_table_name = HistoricalAppropriationAccountBalances._meta.db_table
+                with Timer(f"Empty {destination_table_name}"):
+                    cursor.execute(f"delete from {destination_table_name}")
```

Comment:
since this isn't wrapped in a transaction block and it always performs a full reload, I do prefer `TRUNCATE TABLE <table> RESTART IDENTITY`

Reply:
It is wrapped in a transaction.  It's using psycopg2 which transactions by default.  Django doesn't transaction by default.

Reply:
Interesting.

Even more interesting when I recall Django's connections are on top of psycopg2

Reply:
/shrug

It's been Django's position for a while.

https://docs.djangoproject.com/en/2.2/topics/db/transactions/#autocommit-details